### PR TITLE
feat(remap transform): Add `tag_types_externally` function

### LINKED ISSUE #5727: New `externally_tag_types` Remap function
I'm open to better names for this, and I'm curious if there is a specific term for this, but this function will externally tag all values with its value type. This is particularly useful for storages that require consistent types (like Elasticsearch). This function would ensure that type conflicts never exist. This strategy reminds me of DynamoDB since this is how it avoids typ conflicts.

## Example

For example, given this log event:

```json
{
  "message": "Hello world",
  "request": {
    "duration_ms": 67.9
  }
}
```

And the following Remap expression:

```js
. = externally_tag_types(.)
```

The log event would be transformed into:

```json
{
  "message": {
    "string": "Hello world"
  },
  "request": {
    "duration_ms": {
      "float": 67.9
    }
  }
}
```

### PR DESCRIPTION
**Pull request by pablosichert**
Closes #5727.

### CODE REVIEW

File: tests/behavior/transforms/remap.toml
Comment (pablosichert) [inline tests/behavior/transforms/remap.toml:1288]: This fails when I change it to

```
"regex": /.*/
```

Any ideas why?

File: lib/remap-functions/src/tag_types_externally.rs
Comment (pablosichert) [inline lib/remap-functions/src/tag_types_externally.rs:45]: Is there a way to say "this is an `Array` when `value` is an `Array`, and a `Map` otherwise"?

File: lib/remap-functions/src/tag_types_externally.rs
Comment (pablosichert) [inline lib/remap-functions/src/tag_types_externally.rs:51]: This is still to debate if it shouldn't rather be `"string"`.

**Comment by pablosichert** [comment]
I think it would be reasonable to hold this off until #6353 is merged.

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (JeanMertz) [inline lib/vrl/stdlib/src/tag_types_externally.rs:66]: It's not entirely clear to me why this is here.

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (JeanMertz) [inline lib/vrl/stdlib/src/tag_types_externally.rs:70]: Couldn't this take `Value` by value instead of reference?

**Review by JeanMertz — APPROVED** [review]
Two nits, and it would probably be a good idea to re-enable the tests again in this PR, but other than that, this looks good to go.

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (pablosichert) [reply lib/vrl/stdlib/src/tag_types_externally.rs:66]: That's because of https://github.com/timberio/vector/pull/6382/files#diff-172dbff4177e29056493ffeb3f1ebb10295dfbffb3652f5b4b58beaaf13e283eR167-R176. Array as input also results in an array as output.

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (pablosichert) [reply lib/vrl/stdlib/src/tag_types_externally.rs:70]: Yes, I was also unsure about this. Why would we want this to be consuming when it doesn't need to be, though?

File: lib/remap-functions/src/tag_types_externally.rs
Comment (pablosichert) [reply lib/remap-functions/src/tag_types_externally.rs]: Do you have a tip here, @JeanMertz?

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (JeanMertz) [reply lib/vrl/stdlib/src/tag_types_externally.rs:70]: Well, it _does_ need to be, since you're now cloning the values inside?

But I haven't looked too deep, so there might be a reason somewhere that prevents you from taking ownership here.

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (JeanMertz) [reply lib/vrl/stdlib/src/tag_types_externally.rs:66]: I see, then it would be good to (1) add an `Example` demonstrating this, and (2) update the docs to mention it can also return an array.

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (JeanMertz) [reply lib/vrl/stdlib/src/tag_types_externally.rs:70]: So what I'm proposing here is:

1. take by value
2. remove all the `clone()` calls for scalar values
3. update the collection value's to use `into_iter`

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (JeanMertz) [reply lib/vrl/stdlib/src/tag_types_externally.rs:66]: Also, in that case, shouldn't we add a conditional here, to check the type of `value` and define the resulting type def based on its type def?

That would allow for more precise type definitions when calling this function.

**Comment by JeanMertz** [comment]
@pablosichert shall we get this in before we release 0.12?

**Comment by pablosichert** [comment]
Yes, I'll bring it up to date today 👍🏻

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (pablosichert) [reply lib/vrl/stdlib/src/tag_types_externally.rs]: I confused myself there, thought that consuming the value could mutate nested values, which obviously isn't the case. Thanks for the suggestion!

File: tests/behavior/transforms/remap.toml
Comment (pablosichert) [reply tests/behavior/transforms/remap.toml]: The CLI exposes the behavior I would have expected:

```
$ . = tag_types_externally(r'.*')
{ "regex": /.*/ }

$ .regex == r'.*'
true
```

Still not sure why this test only works with a string literal:

https://github.com/timberio/vector/blob/a497243d6abd8a248ef047f7a7144e4edcdf0e50/tests/behavior/transforms/remap.toml#L1074

Any idea, @JeanMertz?

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (FungusHumungus) [inline lib/vrl/stdlib/src/tag_types_externally.rs:84]: Externally, `bytes` is known as `string`, so we should tag with that instead.

Comment (FungusHumungus) [reply lib/vrl/stdlib/src/tag_types_externally.rs:84]: Oh, I see this was discussed, but not resolved. I still cast my vote towards `string`, you are correct in that it is not guaranteed to be valid utf8, but for all practical purposes I think we can assume it is for this..

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (FungusHumungus) [inline lib/vrl/stdlib/src/tag_types_externally.rs:106]: Likewise from [here](https://github.com/timberio/vector/issues/5727#issuecomment-774510068), nulls should just be removed..

**Comment by FungusHumungus** [comment]
Would you be able to add a benchmark for this to https://github.com/timberio/vector/blob/master/lib/vrl/stdlib/benches/benches.rs?

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (pablosichert) [reply lib/vrl/stdlib/src/tag_types_externally.rs:106]: But in many cases `null` can carry information and would just be missing from the document then?

Or are you proposing to change it to

```suggestion
        Value::Null => (None, Value::Null),
```

Comment (FungusHumungus) [reply lib/vrl/stdlib/src/tag_types_externally.rs:106]: I have no strong opinions either way - just trying to interpret that comment.  Might be worth confirming with @binarylogic.

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (pablosichert) [reply lib/vrl/stdlib/src/tag_types_externally.rs:84]: The issue description has an example with a `"string"` label, so I don't think it implied not to tag it.

Will change it to `"string"` now.

File: lib/remap-functions/src/tag_types_externally.rs
Comment (pablosichert) [reply lib/remap-functions/src/tag_types_externally.rs]: See https://github.com/timberio/vector/pull/6382#discussion_r588219838.

File: lib/vrl/stdlib/src/tag_types_externally.rs
Comment (pablosichert) [reply lib/vrl/stdlib/src/tag_types_externally.rs]: I changed it to the more concise version now (meaning: `null` stays `null`).

My reasoning for keeping the `null`s is: If you want to remove them, you can do that independendly of this transformation. The reverse (adding them back if they are removed by default) would be quite hard.

Comment (binarylogic) [reply lib/vrl/stdlib/src/tag_types_externally.rs]: Agree with @pablosichert, it's fine to keep nulls. The purpose of this feature is to avoid type conflicts for storages that require it. I could see the presence of a field meaning something even if it does not have a value.

File: lib/remap-functions/src/tag_types_externally.rs
Comment (JeanMertz) [reply lib/remap-functions/src/tag_types_externally.rs]: Nothing builder-like right now, no, you'll have to use an if-else branch.

File: tests/behavior/transforms/remap.toml
Comment (JeanMertz) [reply tests/behavior/transforms/remap.toml]: The CLI works entirely with `vrl::Value` types, which has a `VRL::Regex` type. When we interface with Vector, we convert VRL types to Vector types (`vector::Value`).

In Vector, we don't have a regex type because there is no sensible serialization strategy that really makes sense when going to JSON for example.

So, in your CLI example, you're comparing `vrl::Value` types, in the transform test, you're comparing `vector::Value` types, and we convert the regex type to a string representing the actual regex, which is why you're seeing what you're seeing.

At one point we actually returned an error when you tried to send a Regex type from VRL to Vector (basically meaning you could only assign regex types to variables, and couldn't have a type returning from your program that was or contained a regex). I removed this, as it was buggy and caused more trouble than it's worth for now, but we might re-introduce this if this is going to confuse people (I suspect it won't, but I can't be sure).

**Review by JeanMertz — APPROVED** [review]

File: tests/behavior/transforms/remap.toml
Comment (JeanMertz) [reply tests/behavior/transforms/remap.toml]: One additional point.

This appears to be a bug:

```
$ . = tag_types_externally(r'.*')
{ "regex": /.*/ }
```

This should have said:

```
{ "regex": r'.*' }
```

Apparently, we haven't updated our `Display` impl for `Value` yet to switch from `/foo/` to `r'foo'`, so if you're willing to do so, go ahead (I'm okay with seeing that changed either in this PR, or a separate one).

Comment (pablosichert) [reply tests/behavior/transforms/remap.toml]: Opened https://github.com/timberio/vector/pull/6668.

**Pull request merged**
